lib/ansible/modules/cloud/amazon/lambda.py

```diff
@@ -298,15 +298,15 @@ def main():
     except (botocore.exceptions.ClientError, botocore.exceptions.ValidationError) as e:
         module.fail_json(msg=str(e))
 
-    if role.startswith('arn:aws:iam'):
+    if role.startswith('arn:aws:iam') or role.startswith('arn:aws-us-gov:iam'):
```

Comment:
Why not get rid of this if statement altogether and construct the role ARN dynamically all the time?

Reply:
The lambda module is trying to juggle a couple of use cases for role.

1) If a lambda function is being created, a role must be specified.  
2) If the lambda function already exists, the role is optional.   

A straight up `if role is not None` check doesn't satisfy the 2nd use case where the role can be optional on an update.

Additionally the boto3 lambda API requires the role to be the complete ARN. The original author is allowing the shortcut of only providing the role name, however must build the ARN.

It is correct that there could be future partitions.  The current boto3 library does a regex check on the arn and only allows 'aws' or 'aws-us-gov'.  See error message in [#28699](https://github.com/ansible/ansible/issues/28699) while not completely future proof, does make a reasonable assumption based on the boto3 library.  However a more thorough check may be 

```
arn_parts = self.role.split(':')

if len(arn_parts) == 1:
        try:
            iam_client = boto3_conn(module, conn_type='client', resource='iam',
                                    region=region, endpoint=ec2_url, **aws_connect_kwargs)
            arn, partition, service, region, account_id, resource = iam_client.get_user()['User']['Arn'].split(':')
            role_arn = 'arn:{0}:iam::{1}:role/{2}'.format(partition, account_id, role)
        except (botocore.exceptions.ClientError, botocore.exceptions.ValidationError) as e:
            module.fail_json(msg=str(e))
elif len(arn_parts) == 6:
   role_arn = self.role
else:
   module.fail_json(msg="Invalid role ARN")

```

Reply:
I like the approach you stated. I was going to suggest that it could be possible to get the partitions list from boto3 by using boto3's `session.get_available_partitions()` for the `if role.startswith` if you wanted to keep it simple.